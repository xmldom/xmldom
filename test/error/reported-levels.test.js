'use strict'

const { LINE_TO_ERROR_INDEX, REPORTED } = require('./reported')
const { getTestParser } = require('../get-test-parser')
const { ParseError } = require('../../lib/sax')
const { DOMParser } = require('../../lib/dom-parser')

describe.each(Object.entries(REPORTED))(
	'%s',
	(name, { source, level, match, skippedInHtml }) => {
		describe.each(['text/xml', 'text/html'])('with mimeType %s', (mimeType) => {
			const isHtml = mimeType === 'text/html'
			if (isHtml && skippedInHtml) {
				it(`should not be reported as ${level}`, () => {
					const { errors, parser } = getTestParser()

					parser.parseFromString(source, mimeType)

					// if no report was triggered, the key is not present on `errors`
					expect(errors[level]).toBeUndefined()
				})
			} else {
				it(`should be reported as ${level}`, () => {
					const { errors, parser } = getTestParser()

					parser.parseFromString(source, mimeType)

					const reported = errors[level]
					// store the snapshot, so any change in message can be inspected in the git diff
					expect(reported).toMatchSnapshot()
					// if a match has been defined, filter messages
					expect(
						match ? (reported || []).filter(match) : reported
					).toHaveLength(1)
				})
				if (level === 'fatalError') {
					it(`should throw ParseError in errorHandler.fatalError`, () => {
						const parser = new DOMParser()

						expect(() => parser.parseFromString(source, mimeType)).toThrow(
							ParseError
						)
					})
				} else if (level === 'error') {
					it(`should not catch Error thrown in errorHandler.${level}`, () => {
						let thrown = []
						const errorHandler = {
							[level]: jest.fn((message) => {
								const toThrow = new Error(message)
								thrown.push(toThrow)
								throw toThrow
							}),
						}
						const { parser } = getTestParser({ errorHandler })

						expect(() => parser.parseFromString(source, mimeType)).toThrow(
							Error
						)
						expect(
							thrown.map((error) => toErrorSnapshot(error, 'lib/sax.js'))
						).toMatchSnapshot()
						match && expect(match(thrown[0].toString())).toBe(true)
					})
				} else if (level === 'warning') {
					it('should escalate Error thrown in errorHandler.warning to errorHandler.error', () => {
						let thrown = []
						const errorHandler = {
							warning: jest.fn((message) => {
								const toThrow = new Error(message)
								thrown.push(toThrow)
								throw toThrow
							}),
							error: jest.fn(),
						}
						const { parser } = getTestParser({ errorHandler })

						parser.parseFromString(source, mimeType)

						expect(errorHandler.warning).toHaveBeenCalledTimes(1)
						expect(errorHandler.error).toHaveBeenCalledTimes(1)
						expect(
							thrown.map((error) => toErrorSnapshot(error, 'lib/sax.js'))
						).toMatchSnapshot()
						match && expect(match(thrown[0].message)).toBe(true)
					})
				}
			}
		})
	}
)

/**
 * Creates a string from an error that is easily readable in a snapshot
 * - put's the message on one line as first line
 * - picks the first line in the stack trace that is in `libFile`,
 *   and strips absolute paths and character position from that stack entry
 *   as second line. the line number in the stack is converted to the error index
 *   (to make snapshot testing possible even with stryker).
 * @param {Error} error
 * @param {string} libFile the path from the root of the project that should be preserved in the stack
 * @returns {string}
 */
function toErrorSnapshot(error, libFile) {
	const libFileMatch = new RegExp(`\/.*\/(${libFile})`)
	return `${error.message.replace(/([\n\r]+\s*)/g, '||')}\n${error.stack
		.split(/[\n\r]+/)
		// find first line that is from lib/sax.js
		.filter((l) => libFileMatch.test(l))[0]
		// strip of absolute path
		.replace(libFileMatch, '$1')
		// strip of position of character in line
		.replace(/:\d+\)$/, ')')
		// We only store the error index int he snapshot instead of the line numbers.
		// This way they need to be updated less frequent and are compatible with stryker.
		// see `parseErrorLines` in `./reported.js` for how LINE_TO_ERROR_INDEX is created,
		// and `./reported.json` (after running the tests) to inspect it.
		.replace(new RegExp(`${libFile}:\\d+`), (fileAndLine) => {
			return `${libFile}:#${
				fileAndLine in LINE_TO_ERROR_INDEX
					? LINE_TO_ERROR_INDEX[fileAndLine].index
					: -1
			}`
		})}`
}
