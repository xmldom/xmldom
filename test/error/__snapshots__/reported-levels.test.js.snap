// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SYNTAX_AttributeEqualMissingValue with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value missed!!
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_AttributeEqualMissingValue with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value missed!!||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_AttributeEqualMissingValue with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value missed!!
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_AttributeEqualMissingValue with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value missed!!||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_AttributeMissingEndingQuote with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value no end '\\"' match
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_AttributeMissingEndingQuote with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value no end '\\"' match||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_AttributeMissingEndingQuote with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value no end '\\"' match
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_AttributeMissingEndingQuote with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: attribute value no end '\\"' match||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_ElementClosingNotConnected with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: elements closed character '/' and '>' must be connected to
@#[line:1,col:6]",
]
`;

exports[`SYNTAX_ElementClosingNotConnected with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: elements closed character '/' and '>' must be connected to||@#[line:1,col:6]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_ElementClosingNotConnected with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: elements closed character '/' and '>' must be connected to
@#[line:1,col:6]",
]
`;

exports[`SYNTAX_ElementClosingNotConnected with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: elements closed character '/' and '>' must be connected to||@#[line:1,col:6]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_EndTagMaybeNotComplete with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	end tag name: inner maybe not complete
@#[line:1,col:6]",
]
`;

exports[`SYNTAX_EndTagMaybeNotComplete with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	end tag name: inner maybe not complete||@#[line:1,col:6]
    at parse (lib/sax.js:#2)",
  "[xmldom error]	element parse error: Error: [xmldom error]	end tag name: inner maybe not complete||@#[line:1,col:6]||@#[line:1,col:6]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_EndTagMaybeNotComplete with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	end tag name: inner maybe not complete
@#[line:1,col:6]",
]
`;

exports[`SYNTAX_EndTagMaybeNotComplete with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	end tag name: inner maybe not complete||@#[line:1,col:6]
    at parse (lib/sax.js:#2)",
  "[xmldom error]	element parse error: Error: [xmldom error]	end tag name: inner maybe not complete||@#[line:1,col:6]||@#[line:1,col:6]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_EndTagNotComplete with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	end tag name: xml is not complete:xml
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_EndTagNotComplete with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	end tag name: xml is not complete:xml||@#[line:1,col:1]
    at parse (lib/sax.js:#1)",
  "[xmldom error]	element parse error: Error: [xmldom error]	end tag name: xml is not complete:xml||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_EndTagNotComplete with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	end tag name: xml is not complete:xml
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_EndTagNotComplete with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	end tag name: xml is not complete:xml||@#[line:1,col:1]
    at parse (lib/sax.js:#1)",
  "[xmldom error]	element parse error: Error: [xmldom error]	end tag name: xml is not complete:xml||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_InvalidAttributeName with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid attribute:123
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_InvalidAttributeName with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid attribute:123||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_InvalidAttributeName with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid attribute:123
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_InvalidAttributeName with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid attribute:123||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_InvalidTagName with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid tagName:123
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_InvalidTagName with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid tagName:123||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_InvalidTagName with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid tagName:123
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_InvalidTagName with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	element parse error: Error: invalid tagName:123||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_UnclosedComment with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	Unclosed comment
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_UnclosedComment with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	Unclosed comment||@#[line:1,col:1]
    at parseDCC (lib/sax.js:#21)",
  "[xmldom error]	element parse error: Error: [xmldom error]	Unclosed comment||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_UnclosedComment with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	Unclosed comment
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_UnclosedComment with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	Unclosed comment||@#[line:1,col:1]
    at parseDCC (lib/sax.js:#21)",
  "[xmldom error]	element parse error: Error: [xmldom error]	Unclosed comment||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_UnexpectedEndOfInput with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	unexpected end of input
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_UnexpectedEndOfInput with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	unexpected end of input||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#13)",
  "[xmldom error]	element parse error: Error: [xmldom error]	unexpected end of input||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`SYNTAX_UnexpectedEndOfInput with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	unexpected end of input
@#[line:1,col:1]",
]
`;

exports[`SYNTAX_UnexpectedEndOfInput with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	unexpected end of input||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#13)",
  "[xmldom error]	element parse error: Error: [xmldom error]	unexpected end of input||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`WF_AttributeMissingQuote with mimeType text/html should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"value\\" missed quot(\\")!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingQuote with mimeType text/html should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"value\\" missed quot(\\")!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#14)",
]
`;

exports[`WF_AttributeMissingQuote with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"value\\" missed quot(\\")!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingQuote with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"value\\" missed quot(\\")!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#14)",
]
`;

exports[`WF_AttributeMissingQuote2 with mimeType text/html should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"&\\" missed quot(\\")!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingQuote2 with mimeType text/html should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"&\\" missed quot(\\")!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#17)",
]
`;

exports[`WF_AttributeMissingQuote2 with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"&\\" missed quot(\\")!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingQuote2 with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"&\\" missed quot(\\")!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#17)",
]
`;

exports[`WF_AttributeMissingStartingQuote with mimeType text/html should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed start quot(\\")!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingStartingQuote with mimeType text/html should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed start quot(\\")!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#10)",
]
`;

exports[`WF_AttributeMissingStartingQuote with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed start quot(\\")!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingStartingQuote with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed start quot(\\")!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#10)",
]
`;

exports[`WF_AttributeMissingValue with mimeType text/html should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingValue with mimeType text/html should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#15)",
]
`;

exports[`WF_AttributeMissingValue with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingValue with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#15)",
]
`;

exports[`WF_AttributeMissingValue2 with mimeType text/html should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead2!!
@#[line:1,col:1]",
  "[xmldom warning]	attribute \\"attr2\\" missed value!! \\"attr2\\" instead!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingValue2 with mimeType text/html should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead2!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#18)",
]
`;

exports[`WF_AttributeMissingValue2 with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead2!!
@#[line:1,col:1]",
  "[xmldom warning]	attribute \\"attr2\\" missed value!! \\"attr2\\" instead!!
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeMissingValue2 with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute \\"attr\\" missed value!! \\"attr\\" instead2!!||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#18)",
]
`;

exports[`WF_AttributeValueMustAfterEqual with mimeType text/html should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute value must after \\"=\\"
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeValueMustAfterEqual with mimeType text/html should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute value must after \\"=\\"||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#8)",
]
`;

exports[`WF_AttributeValueMustAfterEqual with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	attribute value must after \\"=\\"
@#[line:1,col:1]",
]
`;

exports[`WF_AttributeValueMustAfterEqual with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	attribute value must after \\"=\\"||@#[line:1,col:1]
    at parseElementStartPart (lib/sax.js:#8)",
]
`;

exports[`WF_DuplicateAttribute with mimeType text/html should be reported as fatalError 1`] = `
Array [
  "[xmldom fatalError]	Attribute a redefined
@#[line:1,col:1]",
]
`;

exports[`WF_DuplicateAttribute with mimeType text/xml should be reported as fatalError 1`] = `
Array [
  "[xmldom fatalError]	Attribute a redefined
@#[line:1,col:1]",
]
`;

exports[`WF_EntityDeclared with mimeType text/html should be reported as error 1`] = `
Array [
  "[xmldom error]	entity not found:&e;
@#[line:1,col:1]",
]
`;

exports[`WF_EntityDeclared with mimeType text/html should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	entity not found:&e;||@#[line:1,col:1]
    at replace (lib/sax.js:#0)",
  "[xmldom error]	element parse error: Error: [xmldom error]	entity not found:&e;||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`WF_EntityDeclared with mimeType text/xml should be reported as error 1`] = `
Array [
  "[xmldom error]	entity not found:&e;
@#[line:1,col:1]",
]
`;

exports[`WF_EntityDeclared with mimeType text/xml should not catch Error thrown in errorHandler.error 1`] = `
Array [
  "[xmldom error]	entity not found:&e;||@#[line:1,col:1]
    at replace (lib/sax.js:#0)",
  "[xmldom error]	element parse error: Error: [xmldom error]	entity not found:&e;||@#[line:1,col:1]||@#[line:1,col:1]
    at parse (lib/sax.js:#5)",
]
`;

exports[`WF_UnclosedXmlAttribute with mimeType text/xml should be reported as warning 1`] = `
Array [
  "[xmldom warning]	unclosed xml attribute
@#[line:1,col:1]",
]
`;

exports[`WF_UnclosedXmlAttribute with mimeType text/xml should escalate Error thrown in errorHandler.warning to errorHandler.error 1`] = `
Array [
  "[xmldom warning]	unclosed xml attribute||@#[line:1,col:1]
    at parse (lib/sax.js:#4)",
]
`;
